### DISCUSSION TITLE
Why doesn't get_to() support types that satisfy MoveAssignable but not DefaultConstructible?

### DISCUSSION THREAD

**Discussion by D-Barber** [nlohmann/json#2445] [Q&A]
After looking at the SFINAE checks (and comments) in the `get_to()` implementation I believe that a case I encountered, while it was initially a surprise, is by design. I have been unable to find an explanation of the rationale behind the decision so am interested to understand for certain what it is. I'd also like to know whether the solution I have identified is intentionally supported (and therefore should continue to work in future releases) or works by chance.

I have attached a simplified example of the case below. I essentially replaced a member, which was already being deserialised with `get_to()`, with a non-default-constructible type which already had an `adl_serializer` specialisation defined. This failed to compile because `get_to()` will only use a "method of the form `void from_json(const basic_json&, ValueType&)`". Am I right in thinking that this is in order to avoid `get_to()` ever creating a temporary object of `ValueType` (in order to perform a move-assignment of the referenced value) because this may be unexpected and/or undesirable in the caller's use-case?

In order to try and solve the issue I have added a `from_json()` overload of the correct form (demonstrated by defining the macro `ADD_GET_TO_OVERLOAD` in the example below) which delegates to the MoveConstructible overload and then performs move-assignment of the referenced value. This appears to work but is the resulting specialisation of `adl_serializer` with two `from_json()` overloads intentionally supported or is this likely to result in unexpected behaviour elsewhere in the library?

```C++
#include <nlohmann/json.hpp>

class foo
{
public:
	explicit foo(const int bar) : bar{bar}
	{
	}

	int get_bar() const
	{
		return bar;
	}

private:
	int bar{};
};

namespace nlohmann
{
	template<>
	struct adl_serializer<foo>
	{
		static void to_json(json& j, const foo& v)
		{
			j = v.get_bar();
		}

		static foo from_json(const json& j)
		{
			return foo{j.get<int>()};
		}

#ifdef ADD_GET_TO_OVERLOAD
		static void from_json(const json& j, foo& v)
		{
			v = from_json(j);
		}
#endif
	};
}

int main()
{
	foo v{0};
	nlohmann::json json = 1;
	json.get_to(v);
	return v.get_bar();
}

```

**Comment by nlohmann** [comment]
This is by design. get_to(ValueType& v) already receives an existing object by reference, so it doesn't need ValueType to be default-constructible for get_to() itself — but it is only enabled when adl_serializer<ValueType> provides the 'void from_json(const basic_json&, ValueType&)' overload, detected via the detail::has_from_json trait (include/nlohmann/detail/meta/type_traits.hpp), which is a separate trait from detail::has_non_default_from_json used for the by-value 'ValueType from_json(const basic_json&)' overload that powers get<T>(). The library intentionally does not auto-synthesize a void-overload from a by-value one, because doing so would always construct a temporary and move-assign it, which it leaves as an explicit opt-in rather than an implicit behavior for every type. Your workaround — defining both the by-value static from_json and a void from_json(const json&, ValueType&) that delegates to it — is exactly the supported pattern for enabling get_to() on non-default-constructible types; the two traits are checked independently, so having both overloads in one adl_serializer specialization is safe and intentional, not an accident.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
